controller/linodemachine_controller_helpers.go

```diff
@@ -94,9 +94,93 @@ func (*LinodeMachineReconciler) newCreateConfig(ctx context.Context, machineScop
 		createConfig.RootPass = uuid.NewString()
 	}
 
+	if machineScope.LinodeCluster.Spec.VPCRef != nil {
+		iface, err := r.getVPCInterfaceConfig(ctx, machineScope, createConfig.Interfaces, logger)
+		if err != nil {
+			logger.Error(err, "Failed to get VPC interface config")
+
+			return nil, err
+		}
+		createConfig.Interfaces = append(createConfig.Interfaces, *iface)
+	}
+
 	return createConfig, nil
 }
 
+func buildInstanceAddrs(linodeInstance *linodego.Instance) []clusterv1.MachineAddress {
+	addrs := []clusterv1.MachineAddress{}
+	for _, addr := range linodeInstance.IPv4 {
+		addrType := clusterv1.MachineExternalIP
+		if addr.IsPrivate() {
+			addrType = clusterv1.MachineInternalIP
+		}
+		addrs = append(addrs, clusterv1.MachineAddress{
+			Type:    addrType,
+			Address: addr.String(),
+		})
+	}
+
+	return addrs
+}
+
+func (r *LinodeMachineReconciler) getOwnerMachine(ctx context.Context, linodeMachine infrav1alpha1.LinodeMachine, log logr.Logger) (*clusterv1.Machine, error) {
+	machine, err := kutil.GetOwnerMachine(ctx, r.Client, linodeMachine.ObjectMeta)
+	switch {
+	case err != nil:
+		if err = client.IgnoreNotFound(err); err != nil {
+			log.Error(err, "Failed to fetch owner machine")
+		}
+
+		return nil, err
+	case machine == nil:
+		log.Info("Machine Controller has not yet set OwnerRef, skipping reconciliation")
+
+		return nil, err
+	case skippedMachinePhases[machine.Status.Phase]:
+
+		return nil, err
+	default:
+		match := false
+		for i := range linodeMachine.OwnerReferences {
+			if match = linodeMachine.OwnerReferences[i].UID == machine.UID; match {
+				break
+			}
+		}
+		if !match {
+			log.Info("Failed to find the referenced owner machine, skipping reconciliation", "references", linodeMachine.OwnerReferences, "machine", machine.ObjectMeta)
+
+			return nil, err
+		}
+	}
+
+	return machine, nil
+}
+
+func (r *LinodeMachineReconciler) getClusterFromMetadata(ctx context.Context, machine clusterv1.Machine, log logr.Logger) (*clusterv1.Cluster, error) {
+	cluster, err := kutil.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	switch {
+	case err != nil:
+		if err = client.IgnoreNotFound(err); err != nil {
+			log.Error(err, "Failed to fetch cluster by label")
+		}
+
+		return nil, err
+	case cluster == nil:
+		err = errors.New("missing cluster")
+		log.Error(err, "Missing cluster")
+
+		return nil, err
```

Comment:
The error isn't adding anything useful into the context of the log message.

```suggestion
		log.Error(nil, "Missing cluster")
		return nil, errors.New("missing cluster")
```